Displays email subject, sender, timestamp, and snippet.

pearlzhuzeng opened the pull request:
This pull request renders one `<li>` for each thread in the `INBOX` mailbox. The sender, timestamp, subject, and snippet from the last (most recent) email in each thread are being displayed.
<img width="971" alt="task1" src="https://user-images.githubusercontent.com/19351087/35392068-97e78386-01ae-11e8-8b02-4c0175e84ca2.png">

pearlzhuzeng committed:
Adds helper functions to get sender, timestamp, and snippet

pearlzhuzeng committed:
Modifies data type

pearlzhuzeng committed:
Displays sender, subject, timestamp, and snippet on page

pearlzhuzeng committed:
Extracts duplicates to DRY code

pearlzhuzeng committed:
Makes email address and timecode not displayed

pearlzhuzeng committed:
Puts sender and time formatter into helper functions

pearlzhuzeng committed:
Swaps empty strings with explicit messages

cbothner commented on src/index.js:
```diff
@@ -2,23 +2,50 @@
+  } else {
+    threadIds = [];
+  }
+  return threadIds;
```
It would be a bit more idiomatic to have an “early return”

```js
if (INBOX == null) return []
return INBOX.threadIds
```

cbothner commented on src/index.js:
```diff
@@ -2,23 +2,50 @@
+    threadIds = [];
+  }
+  return threadIds;
+}
```
It would be a bit more idiomatic to have an “early return”

```js
if (INBOX == null) return []
return INBOX.threadIds
```

cbothner commented on src/index.js:
```diff
@@ -2,23 +2,50 @@
+            <button class="email-item" type="button">
+              <div class="sender-details">
+                <p>${formatSender(getMessageSender(message)) ||
+                  'Unknown Sender'}</p>
```
Does it make sense for an email not to have a sender? I think it might be more appropriate to treat this case as an error rather than kindly falling back to a default.

cbothner commented on src/store/index.js:
```diff
@@ -2,16 +2,18 @@
 /**
  * You might find it useful to implement some functions that make accessing
- * deeply nested attributes of Messages or Threads...
+ * deeply nested attributes of Messages or Threads easier...
```
🙏

cbothner commented on src/store/index.js:
```diff
@@ -22,4 +24,31 @@ export function getMessageSubject(message: Message): ?string {
+
+export function getMessageTimestamp(message: Message): ?string {
+  const { headers } = message.payload;
+  const emailTimestamp = headers.find(header => header.name === 'Date');
```
While you won’t have a problem with this sample data, it’s technically not a good idea to trust the Date header since it can be faked. But additionally it’s nice to be able to format the date differently based on locale or time since (i.e. 10:34 AM if it’s today but 10/23/17 if it’s last year) That’s why the API provides `internalDate`

cbothner reviewed (approved):
Awesome work, Pearl. This is very good. Here are just a few thoughts

pearlzhuzeng committed:
DRYs code with early return

pearlzhuzeng committed:
Throws error when there is no sender

pearlzhuzeng committed:
[WIP]Formats time differently
Using localTimeString instead of splitting the string

pearlzhuzeng committed:
Reconcile type
Pull the internalDate string instead of pulling in the Date

pearlzhuzeng committed:
Reconciles type
Moved error throw from the view layer index.js to the helper function layer index.js
